### ISSUE TITLE
Creating a user via CLI without FTP or SSH access

### ISSUE THREAD

**Issue by PitWenkin** [virtualmin/virtualmin-gpl#786]
Until recently (before [7.10.0](https://github.com/virtualmin/virtualmin-gpl/releases/tag/7.10.0)) it was possible to create a user that had no primary email, no FTP and no SSH.

`virtualmin create-user --user test --pass Ab123 --noemail --domain example.com`

This user was only able to access a mail account (`test` in this case)

Now the same command adds FTP and SSH access to the account probalby due to #707?

![Screenshot 2024-04-17 at 10 28 55](https://github.com/virtualmin/virtualmin-gpl/assets/45786226/5db2eabb-499c-4136-bd80-e06d14cb654d)
On the screenshot (`virtual-server/list_users.cgi`) above:

- `sub.demo.sub` was created before the update
- `test` was also created before the update but without the `--noemail` flag, therefore creating an account as well as a email address.
- `test1` was created after the update and was supposed to not have FTP or SSH access.

What would be the correct command to create a user account that ONLY has access to a mail account, without creating an actual email address with the name of the account?

Basically the same thing thats done when user `Add mail user` via the GUI and settings `Primary email address enabled` to `No`, which seems to do what I need:
![Screenshot 2024-04-17 at 10 49 48](https://github.com/virtualmin/virtualmin-gpl/assets/45786226/f114ff80-9a61-46f4-95e3-7ab418ea6ae9)

**Comment by iliajie** [comment]
Hello,

> Now the same command adds FTP and SSH access to the account probalby due to #707?

This is somewhat expected since the Virtualmin CLI always runs as root. To address this issue, you should either always use the `--shell` parameter or change the defaults on the **System Customization ⇾ Custom Shells** page. I recommend explicitly using the `--shell` parameter.

> What would be the correct command to create a user account that ONLY has access to a mail account, without creating an actual email address with the name of the account?

You got it all right, except you're missing the `--shell /dev/null` parameter.

**Comment by iliajie** [comment]
> Adding that seems to create both types of account as before.

I'm not sure I understand what you mean. It's supposed to create only one user record.

> Could you please confirm wether this is a new requirement or if this was always required, but somehow did not break until recently?

Yes, it is a new change. We no longer restrict master administrators from assigning any shell to users; this limitation only applies to normal users now, at least through the UI. However, as a side effect in the CLI, you now need to specify a particular shell explicitly.

Now, I think this is suboptimal. Jamie, what do you think? Should we make the CLI behave as it did before?

**Comment by jcameron** [comment]
> Now, I think this is suboptimal. Jamie, what do you think? Should we make the CLI behave as it did before?

Yes, we shouldn't change the behavior of APIs between versions.

**Comment by PitWenkin** [comment]
> > Adding that seems to create both types of account as before.
> 
> I'm not sure I understand what you mean. It's supposed to create only one user record.

Both with and without primary email enabled.

**Comment by PitWenkin** [comment]
> > Now, I think this is suboptimal. Jamie, what do you think? Should we make the CLI behave as it did before?
> 
> Yes, we shouldn't change the behavior of APIs between versions.

You are thinking about reversing the obligation to use the `--shell` parameter (and possibly releasing an update for that) ?

If so will the script break if the `--shell` parameter is included? 
-> If I change my scripts to play along with the new parameter it would suck if it would break again later on …

**Comment by PitWenkin** [comment]
Is the `--shell` parameter only necessary for `create-user` or also for other command, such as

- `create-simple-alias`
- `delete-alias`
- `delete-user`
- `modify-user`

`list-aliases` and `list-users` do not seem to be affected…?

**Comment by iliajie** [comment]
> If so will the script break if the --shell parameter is included?

I would say that always defining the shell is good practice! It won’t break again; it will just fall back to whatever is configured as the default (work as before).

> Is the `--shell` parameter only necessary for `create-user` or also for other command, such as

It is only related to an actual Unix user. Aliases have no Unix users.

**Comment by PitWenkin** [comment]
> > Is the `--shell` parameter only necessary for `create-user` or also for other command, such as
> 
> It is only related to an actual Unix user. Aliases have no associated Unix users.

`delete-user` and `modify-user` are therefore affected?

**Comment by iliajie** [comment]
> `delete-user` and `modify-user` are therefore affected?

You don't need to set shell for `modify-user` unless you want to change it.

**Comment by jcameron** [comment]
@iliajie will you create a PR to restore the old behavior?

**Comment by iliajie** [comment]
> @iliajie will you create a PR to restore the old behavior?

Yes, https://github.com/virtualmin/virtualmin-gpl/pull/787.

**Comment by PitWenkin** [comment]
> > @iliajie will you create a PR to restore the old behavior?
> 
> Yes, #787.

A first test of this patch seems to make the `--shell` parameter no longer necessary

**Comment by PitWenkin** [comment]
What is the current state of the whole avalanche, of PR's and commits, I've released with this issue? I lost count …

If everything is fixed, could I get a list of patches to apply so I can test this on my end?

**Comment by iliajie** [comment]
No, you should wait for the next Virtualmin release.

**Comment by iliajie** [comment]
@jcameron I have run more tests and everything seems to work just fine!

Though it seems there is still one small issue remaining. If **Permission login** set to **Email only** in **Manage Virtual Server ⇾ Edit Owner Limits** page then in domain owner login it isn't limited to **Email only** but shows **Email only** and **Email and FTP**.

It's worth pointing out that it doesn't show **Email, FTP and SSH**, and doesn't let to save it either.

**Comment by jcameron** [comment]
That's expected ... we only distinguish between allowing SSH or not.

**Comment by iliajie** [comment]
I don't think that's right .. We should limit the login in this case too, I think .. If an admin sets the **Allowed login type** to **Email only**, it implies they not only don’t want domain owners to login, but also create users with **Email and FTP** logins either, for whatever reasons they might have.

That’s what the help tooltip says as well, right?

Otherwise, it’s as if the domain owner can't log in via FTP but can create a user who is able to ..

**Comment by jcameron** [comment]
Ok I'll take a look and see if I can make this work ..

**Comment by iliajie** [comment]
Any progress on this one?

**Comment by jcameron** [comment]
I'm still thinking about it - we never had this functionality in Virtualmin, so I'm not really convinced it is needed.

**Comment by iliajie** [comment]
I think a good reason to have it now is that not selecting login permissions with SSH actually limits SSH. So, logically, not selecting FTP should limit FTP as well.

**Comment by stefan1959** [comment]
Has there been a patch on this, as if I follow docs on this on create-user
https://www.virtualmin.com/docs/development/api-programs/create-user/
virtualmin create-user --domain example.com --user john --pass password --quota 1024 --real 'John Doe'

should only create a user with mail only as stated by docs

''The --ftp option can be used to give the new user an FTP login as well, though by default, a user will only be given an email account.''

But it create a account with ftp and ssh

![image](https://github.com/virtualmin/virtualmin-gpl/assets/2176313/3abf144c-8dd2-4ca0-a94d-653fe49e2962)

**Comment by PitWenkin** [comment]
> Has there been a patch on this, as if I follow docs on this https://www.virtualmin.com/docs/development/api-programs/create-user/ virtualmin create-user --domain example.com --user john --pass password --quota 1024 --real 'John Doe'
> 
I was told to wait for the next virtualmin release:  https://github.com/virtualmin/virtualmin-gpl/issues/786#issuecomment-2099118441

But as of https://github.com/virtualmin/virtualmin-gpl/issues/786#issuecomment-2061626282
 >  you're missing the --shell /dev/null parameter.

Adding that should create the user as written in the docs

**Comment by stefan1959** [comment]
Shouldn't that be added to docs, or will eventually work.

**Comment by iliajie** [comment]
> I'm still thinking about it - we never had this functionality in Virtualmin, so I'm not really convinced it is needed.
>> I think a good reason to have it now is that not selecting login permissions with SSH actually limits SSH. So, logically, not selecting FTP should limit FTP as well.

Jamie, any update on this?

> Shouldn't that be added to the docs, or will it eventually work.

For now you need to use `--shell` param, and with the next Virtualmin  update it will work as before.

**Comment by jcameron** [comment]
> Jamie, any update on this?

No, not yet .. possibly for the next release though.

**Comment by stefan1959** [comment]
Maybe a fix to the docs in the mean time? That's a easy solution :)

**Comment by iliajie** [comment]
> Maybe a fix to the docs in the mean time? That's a easy solution :)

Document what exactly? As far as I remember Virtualmin CLI API already mentions `--shell` parameter, right?

**Comment by stefan1959** [comment]
the example should have --shell to work (at the moment) as stated later on.
![image](https://github.com/virtualmin/virtualmin-gpl/assets/2176313/eaf0cbe9-e765-4861-94a1-72533951f842)

**Comment by iliajie** [comment]
Thanks for the suggestion, but I don't think we should, as it was a bug, and the command line parameters are obvious to use.

Furthermore, we're releasing a new version of Virtualmin pretty soon.

**Comment by PitWenkin** [comment]
So … this should be resolved with [7.20.0](https://github.com/virtualmin/virtualmin-gpl/releases/tag/7.20.0) ?

**Comment by iliajie** [comment]
> So … this should be resolved with [7.20.0](https://github.com/virtualmin/virtualmin-gpl/releases/tag/7.20.0) ?

Yes, this issue will be resolved in Virtualmin 7.20.0, and you won't need to use the `--shell` parameter when creating the default mail user.
